### ISSUE TITLE
Autonumbered list can not be configured to be alphanumeric in LaTeX since Sphinx 1.8.0

### ISSUE THREAD

**Issue by jessetan** [sphinx-doc/sphinx#6511]
**Describe the bug**

Since #5001 was merged to fix #1431, Sphinx supports alphanumeric list items (a,b,c instead of 1,2,3) when this is explicitly present in the RST source.

If automatic numbering is used (with `#.`), lists are automatically numbered with 1,2,3.

Before 1.8.0 it was possible to change the automatic numbering using `enumitem` package. This is no longer possible because each `enumerate` environment defines its own `\theenumi`, `\theenumii` etc.

**To Reproduce**

A piece of RST:

```rst
This is a list with explicit numbers and letters

1.  Explicit number

    a. Explicit letter
    b. Explicit letter

2.  Explicit number

This is a list with implicit numbers and letters

#.  Implicit number

    #. Implicit number or letter?
    #. Implicit number or letter?

#.  Implicit number
```

With the following in the LaTeX preamble:

```latex
\usepackage{enumitem}
\setlist[enumerate,1]{label=\arabic*.}
\setlist[enumerate,2]{label=\alph*.}
```

Produces the following on Sphinx 1.8.0+:

<img width="241" alt="Screen Shot 2019-06-20 at 09 34 15" src="https://user-images.githubusercontent.com/3284111/59835867-cd230c80-934a-11e9-98b5-499916dbc6ca.png">

**Expected behavior**

On Sphinx 1.7.9 and lower, the result respects the numbering type set by `enumitems`:

![Screen Shot 2019-06-20 at 09 58 07](https://user-images.githubusercontent.com/3284111/59835929-e88e1780-934a-11e9-8f77-a7cc82ce64af.png)

Because `\theenumi` etc is set by each `enumerate`, this is hard to work around by hooking into an environment in the preamble.

It would be great if this could be styled using a package like `enumitem`, but a configuration option in Sphinx is also possible.

**Comment by jfbu** [comment]
One possibility is to modify the latex writer to encapsulate the definitions in output latex code in a latex conditional, something like this

```latex
\begin{enumerate}
\ifsphinxenumerateasinrstsource
\def\theenumi...
\fi
...
\end{enumerate}
```

This latex conditional could be set (globally) to false by various means, for example the LaTeX writer could react to a configuration option in conf.py and pass correspondingly an option to sphinx LaTeX package, in order to set this boolean to false.

Setting the conditional could be done by sphinx.sty code automatically on detecting usage of enumitem "at begin document", but usage of enumitem may be for other things, such as working aroundLaTeX limits on nesting levels (without using the Sphinx provided maxlistdepth mechanism which handles that too).

So perhaps rather a conf.py option of the type `latex_enumerate_as_rst` with default True in conf.py. Any opinion? Your problem would then be fixed simply by setting it to false and then add usage of `enumitem` as you indicate.

Or, assuming only LaTeX-proficient users need this, simply mention in doc that there is a LaTeX boolean and that user can use `'preamble'` key of `latex_elements` to turn it to false.

In all cases, usage of raw latex directives can customize behaviour for one specific list.

It might also be worthwile to abstract the LaTeX writer added code into some LaTeX macros with arguments rather than raw `\def...` with no hooks.

**Comment by jessetan** [comment]
From your suggestions, I prefer a LaTeX boolean over a Python option. The user needs to be proficient in LaTeX anyway to add a custom style, so a Latex boolean using `sphinxsetup` should be fine.

One thing I wonder if adding a conditional at the start of every `enumerate` will not make the .tex file too complex to read (by a human for debugging) and slower to process.
The conditional could be placed in Sphinx' LaTeX writer so the .tex  file remains clean, at the cost of a more complex LaTeX writer and a Python configuration option.

**Comment by jfbu** [comment]
@jessetan in the meantime I did PR #6520 which implements with some other names the suggestions from my previous message. It has a Python config bool, which governs a TeX boolean.

Regarding the impact on processing time of the TeX boolean, it is negligible. The TeX boolean ends up hidden in a macro `\sphinxsetlistlabels` which is the only thing showing up in the .tex file. The TeX boolean is hidden from normal users.

> The conditional could be placed in Sphinx' LaTeX writer so the .tex file remains clean, at the cost of a more complex LaTeX writer and a Python configuration option.

The default since 1.8.0 is that we do want to redefine the labelling of nested items ; the Sphinx LaTeX writer currently adds for this a number of `\def...`  to the contents of all enumerate environment. The PR #6520 cleans this up a bit by adding a single LaTeX macro. The alternative would be to patch the enumerate environment itself which would raise potential conflicts with usage of extra LaTeX packages handling lists.

In your use case, the LaTeX writer could indeed avoid cluttering the .tex file as anyhow the extra line with the LaTeX macro `\sphinxsetlistlabels` will do nothing. But the option I chose was to let the conf.py Python option be an interface not to Sphinx LaTeX writer actions, but to the meaning of `\sphinxsetlistlabels` LaTeX code itself.

**Comment by jessetan** [comment]
Thanks, the PR looks like it will address my usecase, although as a user I would expect this to be set using a `sphinxsetup` key rather than a Python option.

**Comment by jfbu** [comment]
@jessetan only simpler... I have removed the Python option, please checkout #6520 update.

**Comment by tk0miya** [comment]
```
This is a list with implicit numbers and letters

#.  Implicit number

    #. Implicit number or letter?
    #. Implicit number or letter?

#.  Implicit number
```
The explanation is not correct. This is a list with implicit numbers and numbers. In reST spec, auto-enumerator is described as following:

>In addition, the auto-enumerator, "#", may be used to automatically enumerate a list. Auto-enumerated lists may begin with explicit enumeration, which sets the sequence. Fully auto-enumerated lists use arabic numerals and begin with 1. (Auto-enumerated lists are new in Docutils 0.3.8.)
>http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#enumerated-lists

Therefore the child list is an enumerated list with arabic numbers. So latest Sphinx renders it as it marked up.


But I feel customizing the list style with `enumitem` conflicts semantics of the notation. So I'm not sure we should support it. TBH, I feel it is a little hacky to me. So -0 to add some options to Sphinx.
IMO, it would be better to override `visit_enumerated_list` handler for LaTeX writer from custom extension.
```
from sphinx.writers.latex import LaTeXTranslator

class CustomTranslator(LaTeXTranslator):
    def visit_enumerated_list(self, node):
        # custom handler for enumitem package
        self.body.append('\\begin{enumerate}\n')
        if 'start' in node:
            self.body.append('\\setcounter{%s}{%d}\n' % (enum, node['start'] - 1))
        if self.table:
            self.table.has_problematic = True


def setup(app):
    app.set_translator('latex', CustomTranslator, override=True)
```

**Comment by jfbu** [comment]
I don't have strong opinion on this. The reST specs themselves are maybe questionable because they consider arabic numerals as default which is not compatible with all languages. For example (but I did not test because my TeXLive is lacking Hebrew support due to space constraints), with Hebrew language, LaTeX hebrew module modifies behaviour of `\arabic` LaTeX macro to produce Hebrew numerals, as far as I know. Thus PDF output is not like reST specs says, already now, for Hebrew (untested).

In the past we have tried to replace hard-coded LaTeX by customizable macros. I would thus propose at any rate to keep that part of #6520 which replaces 

```
        self.body.append('\\def\\the%s{%s{%s}}\n' % (enum, style, enum))
        self.body.append('\\def\\label%s{%s\\the%s %s}\n' %
                         (enum, prefix, enum, suffix))
        self.body.append('\\makeatletter\\def\\p@%s{\\p@%s %s\\the%s %s}\\makeatother\n' %
                         (enumnext, enum, prefix, enum, suffix))
```

by 

```
        self.body.append('\\sphinxsetlistlabels{%s}{%s}{%s}{%s}{%s}%%\n' %
                         (style, enum, enumnext, prefix, suffix))
```

in the LaTeX writer `visit_enumerated_list()` code, with the suitable definition of `\sphinxsetlistlabels` in `sphinx.sty`, like in #6520 but minus the usage of a TeX boolean. This would thus be simple refactoring of current code. And @jessetan could fix his issue by simply `\renewcommand\sphinxsetlistlabels[5]{}` in preamble.

**Comment by tk0miya** [comment]
I also don't have strong opinion to this. Just my vote is -0. At present, I'd not like to support this. But I may change my opinion if enough reasonable.

**Comment by jfbu** [comment]
As alternative to #6520, I propose #6533 which is simply refactoring. But as aftereffect it is possible for @jessetan to fix reported use case.

**Comment by jfbu** [comment]
#6533 has now been merged into 2.0 at 3eeec09dd5d9ea8e45e564b077fedfe6d1262e64 which fixes this.

#6520 was squashed to only keep idea of sphinxsetup boolean, and rebased onto 2.0 as well. I keep temporarily for discussion, but +0 for merging it.

@jessetan do you agree we can close this and discard #6520 or do you see interest in #6520 ?

**Comment by jessetan** [comment]
The changes in #6533 allow me to fix my use case, and I don't see any added benefit in #6520 for me or others. I'm fine with the current situation, thanks for your attention.
